frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -248,6 +248,42 @@ function ReaderHighlight:onTapXPointerSavedHighlight(ges)
     end
 end
 
+function ReaderHighlight:updateHighlight(page, index, side, direction)
+    if self.ui.document.info.has_pages then -- we do this only if it's epub file
+        return
+    end
+
+    local highlight = self.view.highlight.saved[page][index]
+    local highlight_beginning = highlight.pos0
+    local highlight_end = highlight.pos1
+    if side == 0 then -- we move from pos0
+        if direction == 1 then -- move highlight to the right
+            local updated_highlight = self.ui.document:getNextVisibleWordStart(highlight_beginning)
+            if updated_highlight then -- in theory crengine could return nil
+                self.view.highlight.saved[page][index].pos0 = updated_highlight
+            end
+         else -- move highlight to the left
+            local updated_highlight = self.ui.document:getPrevVisibleWordStart(highlight_beginning)
+            if updated_highlight then
+                self.view.highlight.saved[page][index].pos0 = updated_highlight
+            end
+        end
+    else -- we move from pos1
+        if direction == 1 then
+            local updated_highlight = self.ui.document:getNextVisibleWordEnd(highlight_end)
+            if updated_highlight then
+                self.view.highlight.saved[page][index].pos1 = updated_highlight
+            end
+        else
+            local updated_highlight = self.ui.document:getPrevVisibleWordEnd(highlight_end)
+            if updated_highlight then
+                self.view.highlight.saved[page][index].pos1 = updated_highlight
+            end
+        end
+    end
+    UIManager:setDirty(self.dialog, "ui")
+end
+
 function ReaderHighlight:onShowHighlightDialog(page, index)
     self.edit_highlight_dialog = ButtonDialog:new{
         buttons = {
```

Comment:
eg, here, have a `local buttons = { the existing common buttons}` 
After that, `if not self.ui.document.info.has_pages`, `table.insert(buttons, {your new buttons})`
and move after all that:
```
 self.edit_highlight_dialog = ButtonDialog:new{
        buttons = buttons, -- (the ButtonDialog.buttons = your local buttons
}
```